### ISSUE TITLE
Let's encrypt...

### ISSUE THREAD

**Issue by fichtner** [opnsense/plugins#6]
using either:

security/py-letsencrypt -- APACHE20
security/letsencrypt.sh (bash) -- MIT
security/py-acme-tiny -- MIT

https://github.com/kuba/simp_le -- GPLv3 licensed

via: https://forum.opnsense.org/index.php?topic=2319.0

**Comment by fichtner** [comment]
We now have acme-tiny in our eco-system if anyone wants to try...

    # pkg install acme-tiny

**Comment by fraenki** [comment]
@fichtner: Would you mind to add the `security/letskencrypt` port to OPNsense? It looks like a perfect fit with it's main focus being security.

**Comment by fichtner** [comment]
Was already done, package will be in 16.1.16

**Comment by fraenki** [comment]
Looks like `security/letskencrypt`(or `security/acme-client`) can only be used on the LibreSSL flavour of OPNsense. I don't want a different client for every SSL flavour, so I'm going to checkout security/py-acme-tiny instead.

**Comment by fichtner** [comment]
acme-client gained a "static" option now to be usable with OpenSSL, but our build system is not clever enough for this just yet. Just FYI.

**Comment by fichtner** [comment]
@fraenki from the looks of it, acme-client can now be shipped in both LibreSSL and OpenSSL flavours.

**Comment by fraenki** [comment]
@fichtner Yes, just recently installed it on my FreeBSD test VM. I hope to provide a rough concept for how I want to implement LE in OPNsense. Some workarounds are required, but I hope to integrate it with HAProxy, Load balancer (relayd) and the bundled lighttpd. Alpha-quality code should follow soonish. I'm several weeks behind my schedule, but I hope to provide a first alpha version in january.

**Comment by EugenMayer** [comment]
you might really want to look at https://github.com/Neilpang/acme.sh .. i am using it right now on the opnsense instances, it works pretty good.

The current point would be, to reimport the certis when they are issued into opnsense, so more or less a sync between the cert/key folder of acme.sh folder and opnsense configuration.

Using the external script we get: 
 - more or less all modes supported ( dns, .well-known )
 - DNS server integration ( like cloud front )  - its build in

So a lot less code to maintain or rather develop

**Comment by AdSchellevis** [comment]
@EugenMayer can you post some small examples of how this should work? This sounds practical, the first time I looked at integrating let's encrypt there was just too much glue needed to make it functional, if this is easier I might be willing to do the work for this feature.

**Comment by EugenMayer** [comment]
sure

right now on a fresh 17.1 installation, i do
```
cd /root
git clone https://github.com/Neilpang/acme.sh
cd acme.sh
./acme.sh install
```

the latter command just deploys the cron job for acme automatically get new certificates

to get a new certificate using ACME with in my case cloudfront

```
setenv CF_Key <apikey>
setenv CF_Key <mycfmail>

./acme.sh --issue --dns dns_cf --dnssleep 25 -d <mydomain>
```

You can use other modes then --dns or a dns with a API integration ( you can go the manual way ) and all kinds, so very flexible. But when thinking about opnsense, most probably the webserver will not be accessible by the letsencrypt servers, so ACME using DNS sounds like the perfect match

**Comment by fraenki** [comment]
No examples  required, the plugin is almost complete.

**Comment by EugenMayer** [comment]
@fraenki will it have and cloud-dns integration for ACME dns verifcation?

**Comment by fraenki** [comment]
@EugenMayer If you mean _CloudFlare_, then yes.

**Comment by fichtner** [comment]
Some minor things to do:

```
Jan 27 08:55:33	configd.py: [527c12a2-f1dd-46d9-95bd-34e9b440111b] Inline action failed with OPNsense/AcmeClient OPNsense/AcmeClient/lighttpd-acme-challenge.conf 'collections.OrderedDict object' has no attribute 'AcmeClient' at Traceback (most recent call last): File "/usr/local/opnsense/service/modules/processhandler.py", line 505, in execute return ph_inline_actions.execute(self, inline_act_parameters) File "/usr/local/opnsense/service/modules/ph_inline_actions.py", line 52, in execute filenames = tmpl.generate(parameters) File "/usr/local/opnsense/service/modules/template.py", line 309, in generate raise render_exception Exception: OPNsense/AcmeClient OPNsense/AcmeClient/lighttpd-acme-challenge.conf 'collections.OrderedDict object' has no attribute 'AcmeClient'
Jan 27 08:55:33	configd.py: generate template container OPNsense/AcmeClient
Jan 27 08:55:32	configd.py: [527c12a2-f1dd-46d9-95bd-34e9b440111b] generate template OPNsense/AcmeClient
Jan 27 08:54:28	configd.py: [af999a5b-5613-40c4-8fe8-7e0a39aac536] Script action stderr returned "/usr/local/etc/rc.d/acme_http_challenge: Permission denied"
Jan 27 08:54:28	configd.py: [af999a5b-5613-40c4-8fe8-7e0a39aac536] requesting acme_http_challenge status
```

**Comment by fichtner** [comment]
@fraenki ok, that should be all. can you double-check to approve this?

**Comment by syserr0r** [comment]
I am not seeing this plugin available in the WebUI for 16.7.14 -- is that intentional or is there something I need to do to update the plugins list?

**Comment by fichtner** [comment]
@syserr0r It was added to the tree just today ;) since 16.7.14 is EOL, the plugin will only be shipped with 17.1 and up

**Comment by syserr0r** [comment]
Ah the plugin list is set at build time and not pulled from here -- I hadn't realized.
